### PR TITLE
Fix benchmarking for Python 3.9

### PR DESCRIPTION
**Pull request by movermeyer**
There were a number of problems with running the benchmarks with Python 3.9:

* `udatetime` doesn't compile on Python 3.9
* Running `iso8601utils` on Python 3.6 causes issues since `enum34` collides with the built-in enum.
* `zulu v2.0.0` dropped support for Python 3.4. and 3.5
* Our `format_results.py` script couldn't handle the case where a module couldn't run on the latest Python
    * I generally refactored it to be _somewhat_ easier to read.
* `perform_comparison.py` also got an overhaul to be _somewhat_ cleaner.

I've also created a `Dockerfile` that make it easy to get an environment that has all the Python versions installed.

### CODE REVIEW

File: benchmarking/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+    apt install -y python3.6 python3.6-dev python3.6-venv && \
+    apt install -y python3.7 python3.7-dev python3.7-venv && \
+    apt install -y python3.8 python3.8-dev python3.8-venv && \
+    apt install -y python3.9 python3.9-dev python3.9-venv
+
+# Install the other dependencies blamco!
+RUN apt-get install -y git curl gcc build-essential
+
+# Make Python 3.9 the default `python`
+RUN ln -s /usr/lib/python3.9/_sysconfigdata__linux_x86_64-linux-gnu.py  /usr/lib/python3.9/_sysconfigdata__x86_64-linux-gnu.py && \
```
Comment (thomasst) [inline benchmarking/Dockerfile:20]: I got this error when running it:
```
 ---> Running in 1d2d9757bbf9
ln: failed to create symbolic link '/usr/lib/python3.9/_sysconfigdata__x86_64-linux-gnu.py': File exists
The command '/bin/sh -c ln -s /usr/lib/python3.9/_sysconfigdata__linux_x86_64-linux-gnu.py  /usr/lib/python3.9/_sysconfigdata__x86_64-linux-gnu.py &&     update-alternatives --install /usr/bin/python python /usr/bin/python3.9 10' returned a non-zero code: 1
```

**Review by thomasst — COMMENTED** [review]
The py files could use some auto-formatting (e.g. using `black`) and have a max line length, though this doesn't need to be done in this PR.

File: benchmarking/Dockerfile (hunk 1 above)
Comment (movermeyer) [reply benchmarking/Dockerfile]: Seems things have changed since I made this PR, and the `ln` call is no longer needed. 🎉 

Removed.

**Comment by movermeyer** [comment]
> The py files could use some auto-formatting (e.g. using `black`) and have a max line length, though this doesn't need to be done in this PR.

@thomasst Agreed...this code is not the best. I've taken another pass at cleaning things up, but it could probably be made even better yet. 😅 

---

I ran `black` on it, and cherry-picked the changes I liked (`black` and I don't agree on everything). 
Then I ran `pylint`, and `autopep8` and that triggered a lot of refactoring.

---

Anyways, if you are OK with the fact that these are just the benchmark scripts, let's merge it and move forward?

The Dockerfile is fixed, so you should be able to run it again. 🎉

**Comment by thomasst** [comment]
Thanks, I got the following summary now:
```
________________________________________________________________ summary _________________________________________________________________
  py39: commands succeeded
  py38: commands succeeded
  py37: commands succeeded
  py36: commands succeeded
  py35: commands succeeded
ERROR:  py34: InterpreterNotFound: python3.4
  py27: commands succeeded
```

Since 3.4 is already end-of-life (and unlike 2.7 is easy to upgrade to 3.5), we could just drop it.

Otherwise feel free to merge.

**Comment by movermeyer** [comment]
re: Python 3.4

I'll leave it there for now. Generally I'm of the belief that we can keep old versions for as long as they are not a pain to deal with.

In this case, it became a pain to install Python 3.4 inside the Docker container, so I dropped it from the benchmarking Docker container. IMO, it's not a big deal if our benchmarking chart doesn't contain Python 3.4.

If, on the other hand, CircleCI stopped supporting Python 3.4, then our CI would be broken for Python 3.4 which would be enough of a pain to justify dropping it.

(Aside: There are almost [zero downloads](https://pypistats.org/packages/ciso8601) of ciso8601 for Python 3.4)

**Pull request merged**